components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/device/grant/DeviceFlowGrant.java

```diff
@@ -64,32 +67,56 @@ public boolean validateGrant(OAuthTokenReqMessageContext oAuthTokenReqMessageCon
             log.debug("Getting ready to release token for device_code: " + deviceCode);
         }
 
-        DeviceFlowDO deviceFlowDO = DeviceFlowPersistenceFactory.getInstance().getDeviceFlowDAO()
-                .getAuthenticationDetails(deviceCode, clientId);
-        Date date = new Date();
-        deviceStatus = deviceFlowDO.getStatus();
-        deviceFlowDO.setDeviceCode(deviceCode);
-        if (Constants.NOT_EXIST.equals(deviceStatus)) {
-            throw new IdentityOAuth2Exception(DeviceErrorCodes.INVALID_REQUEST, DeviceErrorCodes.INVALID_REQUEST);
+        try {
+            deviceFlowDO = DeviceFlowPersistenceFactory.getInstance().getDeviceFlowDAO()
+                    .getAuthenticationDetails(deviceCode, clientId);
+            deviceStatus = deviceFlowDO.getStatus();
+            deviceFlowDO.setDeviceCode(deviceCode);
+            setLastPollTime(deviceCode);
+        } catch (IdentityOAuth2Exception e) {
+            deviceStatus = e.getMessage();
+            setLastPollTime(deviceCode);
+            handleInvalidRequests(deviceStatus);
         }
+
+        handleInvalidRequests(deviceStatus, deviceFlowDO);
+        if (Constants.AUTHORIZED.equals(deviceStatus)) {
+            DeviceFlowPersistenceFactory.getInstance().getDeviceFlowDAO().setDeviceCodeExpired(deviceCode,
+                    Constants.EXPIRED);
+            setPropertiesForTokenGeneration(oAuthTokenReqMessageContext, deviceFlowDO);
+        }
+        return true;
```

Comment:
earlier we returned true only when device status is ```AUTHORIZED```. Now with this logic, it will return true even thrown IdentityOAuth2Exception not fall into **handleInvalidRequests** method. please check

Reply:
No it will break the flow. If it goes through the catch block , it will execute `handleInvalidRequests(deviceStatus, deviceFlowDO);` and throw exceptions. If it does not go throw the catch block and if it is an invalid case, it will break the flow when executing `handleInvalidRequests(deviceStatus, deviceFlowDO);`

Reply:
But both handleInvalidRequests has if and if else clauses.... so for example if the e.get message is "server.error" then it will pass through catch block right?

Reply:
Thanks for pointing. Fixed.